File: kernel/src/test/scala/net/lshift/diffa/kernel/differencing/DigestBuilderTest.scala
/**
 * Copyright (C) 2010-2011 LShift Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.lshift.diffa.kernel.differencing

import org.junit.Test
import org.junit.Assert._
import net.lshift.diffa.kernel.events.VersionID
import org.joda.time.DateTime
import org.apache.commons.codec.digest.DigestUtils
import net.lshift.diffa.kernel.util.Dates._
import net.lshift.diffa.kernel.participants._
import collection.mutable.HashMap
import collection.immutable.HashSet

/**
 * Test cases for the digest builder.
 */
class DigestBuilderTest {
  val pair = "A-B"
  val categories = new HashMap[String,String]

  def withBucketing(cf:CategoryFunction) = Map("bizDate" -> cf)

  def add(b:DigestBuilder, v:VersionID, d:DateTime, s:String) = b.add(v, Map("bizDate" -> d.toString()), d, s)

  @Test
  def shouldNotBucketIndividualVersions {
    val builder = new DigestBuilder(withBucketing(IndividualCategoryFunction))

    add(builder, VersionID(pair, "id1"), JUL_9_2010_1, "vsn1")
    add(builder, VersionID(pair, "id2"), JUL_9_2010_1, "vsn2")

    assertEquals(
      HashSet(EntityVersion("id1", Seq(JUL_9_2010_1.toString), JUL_9_2010_1, "vsn1"),
              EntityVersion("id2", Seq(JUL_9_2010_1.toString), JUL_9_2010_1, "vsn2")),
      HashSet(builder.versions: _*))
  }

  @Test
  def shouldBucketByDay {
    val builder = new DigestBuilder(withBucketing(DailyCategoryFunction))

    add(builder, VersionID(pair, "id1"), JUL_8_2010_1, "vsn1")
    add(builder, VersionID(pair, "id2"), JUL_8_2010_2, "vsn2")
    add(builder, VersionID(pair, "id3"), JUL_9_2010_1, "vsn3")

    assertEquals(builder.digests, builder.digests)

    assertEquals(
      HashSet(
        // TODO [#2] Should the youngest or lastUpDated date be taken for the comparison?
        AggregateDigest(Seq("2010-07-08"), JUL_8_2010_1, DigestUtils.md5Hex("vsn1" + "vsn2")),
        AggregateDigest(Seq("2010-07-09"), JUL_9_2010_1, DigestUtils.md5Hex("vsn3"))),
      HashSet(builder.digests: _*))
  }

  @Test
  def shouldBucketByMonth {
    val builder = new DigestBuilder(withBucketing(MonthlyCategoryFunction))

    add(builder, VersionID(pair, "id1"), JUL_8_2010_1, "vsn1")
    add(builder, VersionID(pair, "id2"), JUL_8_2010_2, "vsn2")
    add(builder, VersionID(pair, "id3"), JUL_9_2010_1, "vsn3")
    add(builder, VersionID(pair, "id4"), AUG_2_2010_1, "vsn4")

    assertEquals(
      HashSet(
        // TODO [#2] Should the youngest or lastUpDated date be taken for the comparison?
        AggregateDigest(Seq("2010-07"), JUL_8_2010_1, DigestUtils.md5Hex("vsn1" + "vsn2" + "vsn3")),
        AggregateDigest(Seq("2010-08"), AUG_2_2010_1, DigestUtils.md5Hex("vsn4"))),
      HashSet(builder.digests: _*))
  }

  @Test
  def shouldBucketByYear {
    val builder = new DigestBuilder(withBucketing(YearlyCategoryFunction))

    add(builder, VersionID(pair, "id0"), JUN_6_2009_1, "vsn0")
    add(builder, VersionID(pair, "id1"), JUL_8_2010_1, "vsn1")
    add(builder, VersionID(pair, "id2"), JUL_8_2010_2, "vsn2")
    add(builder, VersionID(pair, "id3"), JUL_9_2010_1, "vsn3")
    add(builder, VersionID(pair, "id4"), AUG_2_2010_1, "vsn4")
    add(builder, VersionID(pair, "id5"), JAN_2_2011_1, "vsn5")
    add(builder, VersionID(pair, "id6"), AUG_11_2011_1, "vsn6")

    assertEquals(
      HashSet(
        // TODO [#2] Should the youngest or lastUpDated date be taken for the comparison?
        AggregateDigest(Seq("2009"), JUN_6_2009_1, DigestUtils.md5Hex("vsn0")),
        AggregateDigest(Seq("2010"), JUL_8_2010_1, DigestUtils.md5Hex("vsn1" + "vsn2" + "vsn3" + "vsn4")),
        AggregateDigest(Seq("2011"), JAN_2_2011_1, DigestUtils.md5Hex("vsn5" + "vsn6"))),
      HashSet(builder.digests: _*))
  }

  /**
   * The idea behind this is that a bucket should be immutable after it
   * has been digested, hence adding a new item to the same bucket
   * after it has been digested should result in an error.  
   */
  @Test
  def sealedBuckets = {
    val builder = new DigestBuilder(withBucketing(YearlyCategoryFunction))
    add(builder, VersionID(pair, "id0"), JUN_6_2009_1, "vsn0")
    builder.digests.foreach(_.digest)
    try {
      add(builder, VersionID(pair, "id1"), JUN_6_2009_1, "vsn1")
      fail("Expected to provoke SealedBucketException")
    }
    catch {
      case e:SealedBucketException =>
    }
  }
}

File: participant-support/src/test/java/net/lshift/diffa/participant/scanning/DigestBuilderTest.java
/**
 * Copyright (C) 2010-2011 LShift Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *         http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package net.lshift.diffa.participant.scanning;

import org.apache.commons.codec.digest.DigestUtils;
import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.junit.Test;

import java.util.*;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.fail;

/**
 * Test cases for the digest builder.
 */
public class DigestBuilderTest {
  private static final ScanAggregation bizDateAggregation =
      new DateAggregation("bizDate", DateGranularityEnum.Daily);
  private static final ScanAggregation someStringAggregation =
      new ByNameAggregation("someString");
  private static final List<ScanAggregation> aggregations = Arrays.asList(
      bizDateAggregation, someStringAggregation);

  private static final DateTime JUN_6_2009_1 = new DateTime(2009, 6, 6, 12, 45, 12, 0, DateTimeZone.UTC);
  private static final DateTime JUN_6_2009_2 = new DateTime(2009, 6, 6, 15, 32, 16, 0, DateTimeZone.UTC);
  private static final DateTime JUN_7_2009_1 = new DateTime(2009, 6, 7, 13, 51, 31, 0, DateTimeZone.UTC);

  @Test
  public void shouldReturnEmptyDigestsForNoInput() {
    DigestBuilder builder = new DigestBuilder(aggregations);
    assertEquals(0, builder.toDigests().size());
  }

  @Test
  public void shouldObserveAllAggregationFactors() {
    DigestBuilder builder = new DigestBuilder(aggregations);

    builder.add("id1", createAttrMap(JUN_6_2009_1, "a"), "vsn1");
    builder.add("id2", createAttrMap(JUN_7_2009_1, "b"), "vsn2");
    builder.add("id3", createAttrMap(JUN_6_2009_2, "c"), "vsn3");
    builder.add("id4", createAttrMap(JUN_6_2009_2, "a"), "vsn4");

    assertEquals(
      new HashSet<ScanResultEntry>(Arrays.asList(
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn1" + "vsn4"), createAttrMap("2009-06-06", "a")),
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn2"), createAttrMap("2009-06-07", "b")),
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn3"), createAttrMap("2009-06-06", "c"))
      )),
      new HashSet<ScanResultEntry>(builder.toDigests()));
  }

  @Test
  public void shouldObserveAttributesThatArentAggregationFactors() {
    DigestBuilder builder = new DigestBuilder(Arrays.asList(bizDateAggregation));

    builder.add("id1", createAttrMap(JUN_6_2009_1, "a"), "vsn1");
    builder.add("id2", createAttrMap(JUN_7_2009_1, "b"), "vsn2");
    builder.add("id3", createAttrMap(JUN_6_2009_2, "c"), "vsn3");
    builder.add("id4", createAttrMap(JUN_6_2009_2, "a"), "vsn4");

    assertEquals(
      new HashSet<ScanResultEntry>(Arrays.asList(
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn1" + "vsn4"), createAttrMap("2009-06-06", "a")),
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn2"), createAttrMap("2009-06-07", "b")),
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn3"), createAttrMap("2009-06-06", "c"))
      )),
      new HashSet<ScanResultEntry>(builder.toDigests()));
  }

  /**
   * The idea behind this is that a bucket should be immutable after it
   * has been digested, hence adding a new item to the same bucket
   * after it has been digested should result in an error.
   */
  @Test
  public void bucketsShouldBeSealedAfterQuery() {
    DigestBuilder builder = new DigestBuilder(aggregations);
    builder.add("id0", createAttrMap(JUN_6_2009_1, "a"), "vsn0");

    builder.toDigests();

    try {
      builder.add("id1", createAttrMap(JUN_6_2009_1, "a"), "vsn1");
      fail("Expected to provoke SealedBucketException");
    } catch (SealedBucketException e) {
    }
  }

  @Test
  public void shouldAddViaScanResultEntries() {
    DigestBuilder builder = new DigestBuilder(aggregations);

    builder.add(ScanResultEntry.forEntity("id1", "vsn1", null, createAttrMap(JUN_6_2009_1, "a")));
    builder.add(ScanResultEntry.forEntity("id2", "vsn2", null, createAttrMap(JUN_7_2009_1, "b")));
    builder.add(ScanResultEntry.forEntity("id3", "vsn3", null, createAttrMap(JUN_6_2009_2, "c")));
    builder.add(ScanResultEntry.forEntity("id4", "vsn4", null, createAttrMap(JUN_6_2009_2, "a")));

    assertEquals(
      new HashSet<ScanResultEntry>(Arrays.asList(
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn1" + "vsn4"), createAttrMap("2009-06-06", "a")),
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn2"), createAttrMap("2009-06-07", "b")),
        ScanResultEntry.forAggregate(DigestUtils.md5Hex("vsn3"), createAttrMap("2009-06-06", "c"))
      )),
      new HashSet<ScanResultEntry>(builder.toDigests()));
  }

  @Test(expected = OutOfOrderException.class)
  public void shouldRejectOutOfOrderIds() {
    DigestBuilder builder = new DigestBuilder(aggregations);

    builder.add(ScanResultEntry.forEntity("id2", "vsn2", null, createAttrMap(JUN_7_2009_1, "b")));
    builder.add(ScanResultEntry.forEntity("id1", "vsn1", null, createAttrMap(JUN_6_2009_1, "a")));

  }

  private static Map<String, String> createAttrMap(DateTime bizDate, String ss) {
    return createAttrMap(bizDate.toString(), ss);
  }
  private static Map<String, String> createAttrMap(String bizDate, String ss) {
    Map<String, String> attrs = new HashMap<String, String>();
    attrs.put("bizDate", bizDate.toString());
    attrs.put("someString", ss);
    return attrs;
  }
}
